wip-0019.md

```diff
@@ -0,0 +1,182 @@
+<pre>
+  WIP: WIP-0019
+  Layer: Consensus (hard fork)
+  Title: Random Number Generation (RNG) functionality
+  Authors: Adán SDPC <adan@witnet.foundation>
+  Discussions-To: https://github.com/witnet/witnet-rust/discussions/2058
+  Status: Draft
+  Type: Standards Track
+  Created: 2021-09-06
+  License: BSD-2-Clause
+</pre>
+
+
+## Abstract
+
+A new type of data source is proposed such that leverages the crowd-attestation mechanism of the Witnet protocol
+to enable Random Number Generation (RNG) functionality inside data requests.
+
+## Motivation and rationale
+
+Just in the same way as price feeds are the fundamental oracle-based building block for many DeFi protocols RNGs are
+quickly becoming a keystone building block for non-fungible tokens (NFTs).  
+
+Many use cases are arising in which the smart contracts generating the NFTs are in need of reliable and tamper-proof
+random inputs for their generative algorithms ensuring the "uniqueness" of each piece. That is, they need to randomly
+assign a distinguishable set of scarce attributes or traits to each newly minted NFT unit. 
+
+The case for oracle-powered RNGs expands beyond NFTs, as many other constructs also benefit from random inputs, e.g.
+in the area of gaming, lotteries, gambling, random committee selection, etc.
+
+Due to the robustness and flexibility of the crowd-attestation mechanism of the Witnet protocol, there is a huge
+opportunity to add RNG capabilities to it with very little and self-contained changes.
+
+In particular, the mechanism hereby introduced consists on a new type of data source in which no external HTTP(S)
+servers are queried, and nodes are instead required to secretly commit secret sequences of bytes. The rest of the
+protocol follows in the traditional way, with the only addition of a new reducer function that concatenates the
+revealed values in a deterministic order and hashes the sequence together so as to produce a single, uniform and
+pseudo-random output that will be used as the result of the request.
+
+Additionally, a new deterministic ordering for reveals inside tally transactions is introduced so as to rule out any
+malleability attempt from tally producers. 
+
+Due to the overlap of Witnet's inherent commitment scheme with this last aggregation step, there is no need for the
+committed bytes to be verifiably random. Under the assumptions of stock hashing functions, even if the committers
+collude around crafting their commitments in such a way that the output is favorable to them, as long as 1 single
+committer remains independent and commits a value that is secret to the rest, the final result will hold the desired
+randomness properties.
+
```

Comment:
The other very unlikely attack would be a DoS attack in which a `REVEAL` transaction is "denied" during the  `reveal_extra_rounds` period. For that to happen, an attacker would have to be able to mine 4 blocks consecutively.

Reply:
I included a paragraph stating this more explicitely